File: send-message/java/SendMessage.java
/**
 * SendSeven API - Send Message Example (Java)
 *
 * Demonstrates how to send a text message using the SendSeven API.
 */

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.Properties;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class SendMessage {

    private final String apiToken;
    private final String tenantId;
    private final String apiUrl;
    private final String conversationId;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public SendMessage() {
        // Load configuration from environment or .env file
        loadEnvFile();

        this.apiToken = getEnv("SENDSEVEN_API_TOKEN", null);
        this.tenantId = getEnv("SENDSEVEN_TENANT_ID", null);
        this.apiUrl = getEnv("SENDSEVEN_API_URL", "https://api.sendseven.com/api/v1");
        this.conversationId = getEnv("CONVERSATION_ID", null);

        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = new ObjectMapper();
    }

    private void loadEnvFile() {
        try {
            Properties props = new Properties();
            FileInputStream fis = new FileInputStream(".env");
            props.load(fis);
            fis.close();

            for (String key : props.stringPropertyNames()) {
                if (System.getenv(key) == null) {
                    System.setProperty(key, props.getProperty(key));
                }
            }
        } catch (IOException e) {
            // .env file not found, using environment variables only
        }
    }

    private String getEnv(String key, String defaultValue) {
        String value = System.getenv(key);
        if (value == null) {
            value = System.getProperty(key, defaultValue);
        }
        return value;
    }

    /**
     * Send a text message to a conversation.
     *
     * @param conversationId The UUID of the conversation
     * @param text The message text to send
     * @return The created message as a JsonNode
     * @throws Exception If the API request fails
     */
    public JsonNode sendMessage(String conversationId, String text) throws Exception {
        ObjectNode payload = objectMapper.createObjectNode();
        payload.put("conversation_id", conversationId);
        payload.put("text", text);
        payload.put("message_type", "text");

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiUrl + "/messages"))
                .header("Authorization", "Bearer " + apiToken)
                .header("X-Tenant-ID", tenantId)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(payload)))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() >= 400) {
            throw new RuntimeException("API Error " + response.statusCode() + ": " + response.body());
        }

        return objectMapper.readTree(response.body());
    }

    public void run() {
        // Validate configuration
        if (apiToken == null || apiToken.isEmpty()) {
            System.err.println("Error: SENDSEVEN_API_TOKEN environment variable is required");
            System.exit(1);
        }

        if (tenantId == null || tenantId.isEmpty()) {
            System.err.println("Error: SENDSEVEN_TENANT_ID environment variable is required");
            System.exit(1);
        }

        if (conversationId == null || conversationId.isEmpty()) {
            System.err.println("Error: CONVERSATION_ID environment variable is required");
            System.exit(1);
        }

        System.out.println("Sending message to conversation: " + conversationId);

        try {
            JsonNode message = sendMessage(conversationId, "Hello from the SendSeven Java SDK! ☕");

            System.out.println("Message sent successfully!");
            System.out.println("  ID: " + message.get("id").asText());
            System.out.println("  Status: " + message.get("status").asText());
            System.out.println("  Created at: " + message.get("created_at").asText());
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    public static void main(String[] args) {
        new SendMessage().run();
    }
}

File: send-message/csharp/SendMessage.cs
/**
 * SendSeven API - Send Message Example (C#)
 *
 * Demonstrates how to send a text message using the SendSeven API.
 */

using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

class Program
{
    private static readonly HttpClient httpClient = new HttpClient();

    // Configuration from environment
    private static string ApiToken => Environment.GetEnvironmentVariable("SENDSEVEN_API_TOKEN") ?? "";
    private static string TenantId => Environment.GetEnvironmentVariable("SENDSEVEN_TENANT_ID") ?? "";
    private static string ApiUrl => Environment.GetEnvironmentVariable("SENDSEVEN_API_URL") ?? "https://api.sendseven.com/api/v1";
    private static string ConversationId => Environment.GetEnvironmentVariable("CONVERSATION_ID") ?? "";

    static async Task Main(string[] args)
    {
        // Load .env file if it exists
        LoadEnvFile();

        // Validate configuration
        if (string.IsNullOrEmpty(ApiToken))
        {
            Console.WriteLine("Error: SENDSEVEN_API_TOKEN environment variable is required");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(TenantId))
        {
            Console.WriteLine("Error: SENDSEVEN_TENANT_ID environment variable is required");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(ConversationId))
        {
            Console.WriteLine("Error: CONVERSATION_ID environment variable is required");
            Environment.Exit(1);
        }

        Console.WriteLine($"Sending message to conversation: {ConversationId}");

        try
        {
            var message = await SendMessageAsync(ConversationId, "Hello from the SendSeven C# SDK! 💜");

            Console.WriteLine("Message sent successfully!");
            Console.WriteLine($"  ID: {message.GetProperty("id").GetString()}");
            Console.WriteLine($"  Status: {message.GetProperty("status").GetString()}");
            Console.WriteLine($"  Created at: {message.GetProperty("created_at").GetString()}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }

    /// <summary>
    /// Send a text message to a conversation.
    /// The recipient is auto-resolved from the conversation's contact method.
    /// No need to specify 'to' when replying to an existing conversation.
    /// </summary>
    private static async Task<JsonElement> SendMessageAsync(string conversationId, string text)
    {
        var payload = new
        {
            conversation_id = conversationId,
            text = text,
            message_type = "text"
        };

        return await SendPayloadAsync(payload);
    }

    /// <summary>
    /// Send a message using a contact method ID.
    /// The contact_method_id resolves the recipient, channel, and contact
    /// automatically. This is the cleanest way to initiate a new message
    /// without needing a conversation_id.
    /// </summary>
    private static async Task<JsonElement> SendMessageViaContactMethodAsync(string contactMethodId, string text)
    {
        var payload = new
        {
            contact_method_id = contactMethodId,
            text = text,
            message_type = "text"
        };

        return await SendPayloadAsync(payload);
    }

    /// <summary>
    /// Send a payload to the messages API endpoint.
    /// </summary>
    private static async Task<JsonElement> SendPayloadAsync(object payload)
    {
        var json = JsonSerializer.Serialize(payload);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        httpClient.DefaultRequestHeaders.Clear();
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
        httpClient.DefaultRequestHeaders.Add("X-Tenant-ID", TenantId);

        var response = await httpClient.PostAsync($"{ApiUrl}/messages", content);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Load environment variables from .env file if it exists.
    /// </summary>
    private static void LoadEnvFile()
    {
        var envPath = Path.Combine(Directory.GetCurrentDirectory(), ".env");
        if (!File.Exists(envPath)) return;

        foreach (var line in File.ReadAllLines(envPath))
        {
            var trimmedLine = line.Trim();
            if (string.IsNullOrEmpty(trimmedLine) || trimmedLine.StartsWith("#")) continue;

            var separatorIndex = trimmedLine.IndexOf('=');
            if (separatorIndex <= 0) continue;

            var key = trimmedLine.Substring(0, separatorIndex).Trim();
            var value = trimmedLine.Substring(separatorIndex + 1).Trim();

            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key)))
            {
                Environment.SetEnvironmentVariable(key, value);
            }
        }
    }
}
